File: frontend/js/views/car_tco.js
// Vehicle Total Cost of Ownership calculator. Depreciation +
// financing interest + fuel + insurance + maintenance (5%/yr
// inflator) + registration − residual value. Reports per-year +
// totals + cost-per-mile.

import { api } from '../api.js';
import { esc } from '../util.js';
import { t } from '../i18n.js';

export async function renderCarTco(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.car_tco.title">// VEHICLE TCO</span></h1>
        <p class="muted small" data-i18n-html="view.car_tco.intro">
            Total cost of ownership over the years you hold the car. Sums depreciation
            (purchase + tax − residual), financing interest, fuel, insurance, maintenance
            (5%/yr inflator), registration. Reports total, cost-per-mile, and a per-year
            breakdown for budgeting.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(200px,1fr));gap:8px;margin-bottom:12px">
                <label><span class="muted small" data-i18n="view.car_tco.field.price">Purchase price $</span>
                    <input type="number" id="ct-price" step="500" min="0" value="35000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.down">Down payment $</span>
                    <input type="number" id="ct-down" step="500" min="0" value="5000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.tax">Sales tax %</span>
                    <input type="number" id="ct-tax" step="0.25" min="0" max="30" value="8" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.apr">APR %</span>
                    <input type="number" id="ct-apr" step="0.25" min="0" max="50" value="6.5" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.term">Loan term months</span>
                    <input type="number" id="ct-term" step="12" min="0" max="120" value="60" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.hold">Hold years</span>
                    <input type="number" id="ct-hold" step="1" min="1" max="30" value="7" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.miles">Annual miles</span>
                    <input type="number" id="ct-miles" step="500" min="0" max="200000" value="12000" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.mpg">MPG</span>
                    <input type="number" id="ct-mpg" step="0.5" min="0" value="28" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.fuel">Fuel $ / gallon</span>
                    <input type="number" id="ct-fuel" step="0.1" min="0" value="3.50" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.insurance">Insurance $/yr</span>
                    <input type="number" id="ct-insurance" step="50" min="0" value="1500" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.maintenance">Maintenance $/yr (yr 1)</span>
                    <input type="number" id="ct-maint" step="50" min="0" value="800" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.registration">Registration $/yr</span>
                    <input type="number" id="ct-reg" step="25" min="0" value="200" style="width:100%"></label>
                <label><span class="muted small" data-i18n="view.car_tco.field.residual">Residual %</span>
                    <input type="number" id="ct-res" step="5" min="0" max="100" value="30" style="width:100%"></label>
            </div>
            <button class="btn btn-sm primary" id="ct-run" data-shortcut="r" data-i18n="view.car_tco.btn.run">⚡ Compute TCO</button>
            <div id="ct-result"></div>
        </div>
    `;
    mount.querySelector('#ct-run').addEventListener('click', () => runCompute(mount));
    await runCompute(mount);
}

async function runCompute(mount) {
    const result = mount.querySelector('#ct-result');
    const input = {
        purchase_price_usd: parseFloat(mount.querySelector('#ct-price').value) || 0,
        down_payment_usd: parseFloat(mount.querySelector('#ct-down').value) || 0,
        sales_tax_pct: parseFloat(mount.querySelector('#ct-tax').value) || 0,
        apr_pct: parseFloat(mount.querySelector('#ct-apr').value) || 0,
        loan_term_months: parseInt(mount.querySelector('#ct-term').value, 10) || 0,
        hold_years: parseInt(mount.querySelector('#ct-hold').value, 10) || 7,
        annual_miles: parseInt(mount.querySelector('#ct-miles').value, 10) || 0,
        mpg: parseFloat(mount.querySelector('#ct-mpg').value) || 0,
        fuel_price_per_gallon_usd: parseFloat(mount.querySelector('#ct-fuel').value) || 0,
        insurance_annual_usd: parseFloat(mount.querySelector('#ct-insurance').value) || 0,
        maintenance_annual_usd: parseFloat(mount.querySelector('#ct-maint').value) || 0,
        registration_annual_usd: parseFloat(mount.querySelector('#ct-reg').value) || 0,
        residual_pct_after_hold: parseFloat(mount.querySelector('#ct-res').value) || 0,
    };
    result.innerHTML = `<p class="muted">${esc(t('view.car_tco.status.computing'))}</p>`;
    try {
        const r = await api.request('/car-tco/compute', { method: 'POST', body: JSON.stringify(input) });
        result.innerHTML = `
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(180px,1fr));gap:12px;margin-top:1rem">
                <div><div class="muted small">${esc(t('view.car_tco.field.total'))}</div>
                    <strong style="font-size:1.4em">$${r.total_cost_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.car_tco.field.cpm'))}</div>
                    <strong>$${r.cost_per_mile_usd.toFixed(3)}/mi</strong></div>
                <div><div class="muted small">${esc(t('view.car_tco.field.depreciation'))}</div>
                    <strong class="neg">$${r.depreciation_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.car_tco.field.residual'))}</div>
                    <strong>$${r.residual_value_usd.toFixed(0)}</strong></div>
                <div><div class="muted small">${esc(t('view.car_tco.field.monthly'))}</div>
                    <strong>$${r.monthly_payment_usd.toFixed(2)}/mo</strong></div>
                <div><div class="muted small">${esc(t('view.car_tco.field.interest'))}</div>
                    <strong class="neg">$${r.total_financing_interest_usd.toFixed(0)}</strong></div>
            </div>
            <h2 style="margin-top:1rem">${esc(t('view.car_tco.h2.totals'))}</h2>
            <table class="trades">
                <tbody>
                    <tr><td><strong>${esc(t('view.car_tco.row.fuel'))}</strong></td><td>$${r.total_fuel_usd.toFixed(0)}</td></tr>
                    <tr><td><strong>${esc(t('view.car_tco.row.insurance'))}</strong></td><td>$${r.total_insurance_usd.toFixed(0)}</td></tr>
                    <tr><td><strong>${esc(t('view.car_tco.row.maintenance'))}</strong></td><td>$${r.total_maintenance_usd.toFixed(0)}</td></tr>
                    <tr><td><strong>${esc(t('view.car_tco.row.registration'))}</strong></td><td>$${r.total_registration_usd.toFixed(0)}</td></tr>
                </tbody>
            </table>
            <h2 style="margin-top:1rem">${esc(t('view.car_tco.h2.yearly'))}</h2>
            <table class="trades">
                <thead><tr>
                    <th data-i18n="view.car_tco.th.year">Yr</th>
                    <th data-i18n="view.car_tco.th.financing">Financing</th>
                    <th data-i18n="view.car_tco.th.fuel">Fuel</th>
                    <th data-i18n="view.car_tco.th.ins">Insurance</th>
                    <th data-i18n="view.car_tco.th.maint">Maintenance</th>
                    <th data-i18n="view.car_tco.th.reg">Registration</th>
                    <th data-i18n="view.car_tco.th.total_yr">Year total</th>
                </tr></thead>
                <tbody>${(r.yearly || []).map(y => `
                    <tr>
                        <td>${y.year}</td>
                        <td>$${y.financing_usd.toFixed(0)}</td>
                        <td>$${y.fuel_usd.toFixed(0)}</td>
                        <td>$${y.insurance_usd.toFixed(0)}</td>
                        <td>$${y.maintenance_usd.toFixed(0)}</td>
                        <td>$${y.registration_usd.toFixed(0)}</td>
                        <td><strong>$${y.total_year_usd.toFixed(0)}</strong></td>
                    </tr>
                `).join('')}</tbody>
            </table>
        `;
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

File: crates/traderview-db/src/car_tco.rs
//! Vehicle Total Cost of Ownership (TCO) calculator.
//!
//! Multi-year TCO model for a vehicle purchase: sums depreciation,
//! financing interest, fuel, insurance, maintenance, registration/fees,
//! and reduces by residual value at sale. Reports year-by-year + total +
//! cost-per-mile.
//!
//! Inputs:
//!   - purchase_price_usd, down_payment_usd, sales_tax_pct
//!   - apr_pct, loan_term_months
//!   - hold_years          — how long you keep the car
//!   - annual_miles
//!   - mpg                  — fuel economy
//!   - fuel_price_per_gallon_usd
//!   - insurance_annual_usd
//!   - maintenance_annual_usd  (starts low, grows with age — we apply
//!     a 5%/yr inflator)
//!   - registration_annual_usd
//!   - residual_pct_after_hold — assumed sale value as % of MSRP
//!     (typical 5-year used-vehicle retention ~50-60%)
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

const MAINTENANCE_INFLATOR_PCT: f64 = 5.0;

#[derive(Debug, Clone, Deserialize)]
pub struct CarTcoInput {
    pub purchase_price_usd: f64,
    #[serde(default)]
    pub down_payment_usd: f64,
    #[serde(default)]
    pub sales_tax_pct: f64,
    #[serde(default)]
    pub apr_pct: f64,
    #[serde(default = "default_term")]
    pub loan_term_months: u32,
    pub hold_years: u32,
    pub annual_miles: u32,
    #[serde(default = "default_mpg")]
    pub mpg: f64,
    #[serde(default = "default_fuel")]
    pub fuel_price_per_gallon_usd: f64,
    #[serde(default = "default_insurance")]
    pub insurance_annual_usd: f64,
    #[serde(default = "default_maintenance")]
    pub maintenance_annual_usd: f64,
    #[serde(default = "default_registration")]
    pub registration_annual_usd: f64,
    #[serde(default = "default_residual")]
    pub residual_pct_after_hold: f64,
}

fn default_term() -> u32 { 60 }
fn default_mpg() -> f64 { 28.0 }
fn default_fuel() -> f64 { 3.50 }
fn default_insurance() -> f64 { 1500.0 }
fn default_maintenance() -> f64 { 800.0 }
fn default_registration() -> f64 { 200.0 }
fn default_residual() -> f64 { 50.0 }

#[derive(Debug, Clone, Serialize)]
pub struct YearRow {
    pub year: u32,
    pub financing_usd: f64,
    pub fuel_usd: f64,
    pub insurance_usd: f64,
    pub maintenance_usd: f64,
    pub registration_usd: f64,
    pub total_year_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct CarTcoReport {
    pub principal_financed_usd: f64,
    pub monthly_payment_usd: f64,
    pub total_financing_interest_usd: f64,
    pub total_fuel_usd: f64,
    pub total_insurance_usd: f64,
    pub total_maintenance_usd: f64,
    pub total_registration_usd: f64,
    pub residual_value_usd: f64,
    pub depreciation_usd: f64,
    pub total_cost_usd: f64,
    pub total_miles: u32,
    pub cost_per_mile_usd: f64,
    pub yearly: Vec<YearRow>,
}

// ─── Pure compute ─────────────────────────────────────────────────────────

pub fn monthly_payment(principal: f64, apr_pct: f64, term_months: u32) -> f64 {
    if term_months == 0 || principal <= 0.0 {
        return 0.0;
    }
    let r = apr_pct / 100.0 / 12.0;
    if r.abs() < 1e-12 {
        return principal / term_months as f64;
    }
    let n = term_months as f64;
    principal * r / (1.0 - (1.0 + r).powf(-n))
}

/// Sum of monthly interest charges over the first `months_to_count`
/// months of a standard amortization. Returns 0 for degenerate inputs
/// (non-positive monthly payment, zero principal). Handles rate=0 as a
/// special case (interest is 0 every month).
fn amortized_interest_for_months(
    principal: f64,
    apr_pct: f64,
    monthly: f64,
    months_to_count: u32,
) -> f64 {
    if principal <= 0.0 || monthly <= 0.0 || months_to_count == 0 {
        return 0.0;
    }
    let r = apr_pct / 100.0 / 12.0;
    if r == 0.0 {
        return 0.0;
    }
    let mut balance = principal;
    let mut interest_total = 0.0;
    for _ in 0..months_to_count {
        if balance <= 0.0 {
            break;
        }
        let interest = balance * r;
        let principal_portion = (monthly - interest).max(0.0);
        let paid_principal = principal_portion.min(balance);
        interest_total += interest;
        balance -= paid_principal;
    }
    interest_total
}

pub fn compute(input: &CarTcoInput) -> CarTcoReport {
    let tax = input.purchase_price_usd * input.sales_tax_pct / 100.0;
    let principal = (input.purchase_price_usd + tax - input.down_payment_usd).max(0.0);
    let monthly = monthly_payment(principal, input.apr_pct, input.loan_term_months);
    let hold = input.hold_years.max(1);
    let mut yearly: Vec<YearRow> = Vec::with_capacity(hold as usize);
    let mut total_fuel = 0.0;
    let mut total_insurance = 0.0;
    let mut total_maintenance = 0.0;
    let mut total_registration = 0.0;
    for y in 1..=hold {
        // Financing for the year — sum 12 monthly payments, but only
        // count payments through end of loan term.
        let months_into_year = ((y - 1) * 12 + 12).min(input.loan_term_months);
        let months_prev_year = ((y - 1) * 12).min(input.loan_term_months);
        let financing_year = (months_into_year - months_prev_year) as f64 * monthly;

        let fuel_year = if input.mpg > 0.0 {
            input.annual_miles as f64 / input.mpg * input.fuel_price_per_gallon_usd
        } else { 0.0 };
        let insurance_year = input.insurance_annual_usd;
        let maint_year = input.maintenance_annual_usd
            * (1.0_f64 + MAINTENANCE_INFLATOR_PCT / 100.0).powi((y - 1) as i32);
        let reg_year = input.registration_annual_usd;
        total_fuel += fuel_year;
        total_insurance += insurance_year;
        total_maintenance += maint_year;
        total_registration += reg_year;
        yearly.push(YearRow {
            year: y,
            financing_usd: financing_year,
            fuel_usd: fuel_year,
            insurance_usd: insurance_year,
            maintenance_usd: maint_year,
            registration_usd: reg_year,
            total_year_usd: financing_year + fuel_year + insurance_year + maint_year + reg_year,
        });
    }
    // Real interest paid during the hold = sum of each month's
    // interest portion from a proper amortization schedule, capped at
    // hold*12 months. The previous formula was
    // `(total_payments - principal).max(0)` which silently CLAMPED to
    // 0 whenever hold < loan_term — because by month `hold*12` the
    // borrower has paid only `hold*12 * monthly` < principal, so the
    // subtraction goes negative and gets zeroed. A 3-year hold of a
    // 5-year auto loan reported $0 financing interest under the old
    // math, even though ~3 years of monthly interest accrued.
    let total_financing_interest =
        amortized_interest_for_months(principal, input.apr_pct, monthly, hold * 12);
    let residual = input.purchase_price_usd * input.residual_pct_after_hold / 100.0;
    let depreciation = (input.purchase_price_usd + tax - residual).max(0.0);
    // Total cost = depreciation + interest paid + fuel + insurance + maintenance + registration
    let total_cost = depreciation + total_financing_interest + total_fuel + total_insurance
        + total_maintenance + total_registration;
    let total_miles = input.annual_miles.saturating_mul(hold);
    let cost_per_mile = if total_miles > 0 { total_cost / total_miles as f64 } else { 0.0 };
    CarTcoReport {
        principal_financed_usd: principal,
        monthly_payment_usd: monthly,
        total_financing_interest_usd: total_financing_interest,
        total_fuel_usd: total_fuel,
        total_insurance_usd: total_insurance,
        total_maintenance_usd: total_maintenance,
        total_registration_usd: total_registration,
        residual_value_usd: residual,
        depreciation_usd: depreciation,
        total_cost_usd: total_cost,
        total_miles,
        cost_per_mile_usd: cost_per_mile,
        yearly,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input() -> CarTcoInput {
        CarTcoInput {
            purchase_price_usd: 35_000.0,
            down_payment_usd: 5_000.0,
            sales_tax_pct: 8.0,
            apr_pct: 6.5,
            loan_term_months: 60,
            hold_years: 7,
            annual_miles: 12_000,
            mpg: 28.0,
            fuel_price_per_gallon_usd: 3.50,
            insurance_annual_usd: 1_500.0,
            maintenance_annual_usd: 800.0,
            registration_annual_usd: 200.0,
            residual_pct_after_hold: 30.0,
        }
    }

    #[test]
    fn monthly_payment_known() {
        let p = monthly_payment(30_000.0, 6.5, 60);
        assert!((p - 586.78).abs() < 5.0);
    }

    #[test]
    fn monthly_payment_zero_apr() {
        assert_eq!(monthly_payment(36_000.0, 0.0, 60), 600.0);
    }

    #[test]
    fn compute_yearly_count_matches_hold() {
        let r = compute(&input());
        assert_eq!(r.yearly.len(), 7);
    }

    #[test]
    fn compute_residual_basic() {
        let r = compute(&input());
        // residual = $35k × 0.30 = $10.5k
        assert_eq!(r.residual_value_usd, 10_500.0);
    }

    #[test]
    fn compute_depreciation_includes_tax() {
        let r = compute(&input());
        // depreciation = (35k + 8% tax) − 10.5k = 37.8k − 10.5k = 27.3k
        let expected = 35_000.0 + 2_800.0 - 10_500.0;
        assert!((r.depreciation_usd - expected).abs() < 0.01);
    }

    #[test]
    fn compute_fuel_basic() {
        let r = compute(&input());
        // 12k mi / 28 mpg × $3.50 × 7 years = $10,500
        let expected = 12_000.0 / 28.0 * 3.50 * 7.0;
        assert!((r.total_fuel_usd - expected).abs() < 1.0);
    }

    #[test]
    fn compute_insurance_summed() {
        let r = compute(&input());
        assert_eq!(r.total_insurance_usd, 1_500.0 * 7.0);
    }

    #[test]
    fn compute_maintenance_inflates() {
        let r = compute(&input());
        // First year = 800, last (year 7) = 800 × 1.05^6 ≈ 1072
        assert_eq!(r.yearly[0].maintenance_usd, 800.0);
        let last = r.yearly.last().unwrap().maintenance_usd;
        assert!(last > 800.0);
        assert!((last - 800.0 * 1.05_f64.powi(6)).abs() < 0.5);
    }

    #[test]
    fn compute_financing_in_year_capped_at_term() {
        let r = compute(&input());
        // hold = 7 years (84 months), loan = 60 months. Years 6 + 7 should
        // have zero or partial financing.
        assert!(r.yearly[5].financing_usd >= 0.0);
        assert_eq!(r.yearly[6].financing_usd, 0.0);
    }

    #[test]
    fn compute_zero_mpg_zero_fuel() {
        let mut i = input();
        i.mpg = 0.0;
        let r = compute(&i);
        assert_eq!(r.total_fuel_usd, 0.0);
    }

    #[test]
    fn compute_cost_per_mile_positive() {
        let r = compute(&input());
        assert!(r.cost_per_mile_usd > 0.0);
        assert!(r.cost_per_mile_usd < 10.0); // sanity bound
    }

    #[test]
    fn compute_zero_miles_zero_cpm() {
        let mut i = input();
        i.annual_miles = 0;
        let r = compute(&i);
        assert_eq!(r.cost_per_mile_usd, 0.0);
    }

    #[test]
    fn compute_total_cost_includes_all_components() {
        let r = compute(&input());
        let sum = r.depreciation_usd + r.total_financing_interest_usd + r.total_fuel_usd
            + r.total_insurance_usd + r.total_maintenance_usd + r.total_registration_usd;
        assert!((r.total_cost_usd - sum).abs() < 0.01);
    }
}
